packages/core/src/api/app/FeatureFlags.tsx

```diff
@@ -0,0 +1,222 @@
+      );
+    }
+
+    window.localStorage.setItem(
+      this.localStorageKey,
+      JSON.stringify(
+        [...flags].reduce((list, flag) => ({ ...list, [flag]: true }), {}),
+      ),
+    );
+  }
+
+  getEnabledFeatureFlags(): Set<FeatureFlagName> {
+    if (!('localStorage' in window)) {
+      throw new Error(
+        'Feature Flags are not supported on browsers without the Local Storage API',
+      );
+    }
+
+    try {
+      const featureFlagsJson = window.localStorage.getItem(
+        this.localStorageKey,
+      );
+      return new Set<FeatureFlagName>(
+        Object.keys(JSON.parse(featureFlagsJson!)),
+      );
+    } catch (err) {
+      return new Set<FeatureFlagName>();
+    }
+  }
+
+  // We don't make this private as we need this to validate
+  // in the `registerFeatureFlag` method in the Plugin API.
+  checkFeatureFlagNameErrors(name: FeatureFlagName): string[] {
+    const errors: string[] = [];
+
+    if (name.length < 3) {
+      errors.push('The name must have a minimum length of three characters.');
+    }
+
+    if (name.length > 150) {
+      errors.push('The name must not exceed 150 characters.');
+    }
+
+    if (!name.match(/^[a-z]+[a-z0-9-]+$/)) {
+      errors.push(
+        'The name must start with a lowercase letter and only contain lowercase letters, numbers and hyphens. ' +
+          'Examples: feature-flag-one, alpha, release-2020',
+      );
+    }
+
+    return errors;
+  }
+
+  useFeatureFlag(
+    name: FeatureFlagName,
+  ): [FeatureFlagState, (state: FeatureFlagState) => void] {
+    // Check for context
+    const context = useContext(FeatureFlagsContext);
+    if (!context) {
+      throw new Error(
+        'No FeatureFlagsContext found. ' +
+          'Please use this React Hook in the context of your <App />',
+      );
+    }
+
+    // Check for errors
+    // eslint-disable-next-line no-use-before-define
+    const errors = FeatureFlags.checkFeatureFlagNameErrors(name);
+    if (errors.length > 0) {
+      throw new Error(errors[0]);
+    }
+
+    // Check if the feature flag is registered
+    const allFlagNames = [...context.featureFlags].map(flag => flag.name);
+    if (!allFlagNames.includes(name)) {
+      throw new Error(
+        `The '${name}' feature flag is not registered by any plugin. ` +
+          `See the 'registerFeatureFlag' method in the Plugin API (or in your plugin.ts file) on how to register Feature Flags.`,
+      );
+    }
+
+    // eslint-disable-next-line no-use-before-define
+    const currentState = FeatureFlags.get(context.enabledFeatureFlags, name);
+    const setState = (state: FeatureFlagState): void => {
+      // Set the value
+      // eslint-disable-next-line no-use-before-define
+      FeatureFlags.set(name, state);
+
+      // Now update the global state
+      context.refreshEnabledFeatureFlags();
+    };
+
+    return [currentState, setState];
+  }
+}
+
+export const FeatureFlags = new FeatureFlagsImpl();
```

Comment:
Export the constructor instead of an instance, the construct it in the app. These type of globals lead to some pretty annoying dependency/testing issues.

Reply:
Can you share more (an article maybe) on what these dependency/testing issues are? I don't see how this is different to exporting an object with methods. Might be wrong but not sure I'm sold on switching `FeatureFlags.{method}` with `new FeatureFlags().{method}` scattered around the codebase

Reply:
I think this answer sums it up, https://stackoverflow.com/questions/484635/are-global-variables-bad

I don't mean that you should be calling `new FeatureFlags().method`, that's what the `ApiProvider` is for. You provide an implementation of the API in the app, and then access that implementation through the react context.

Reply:
Fair enough. Fixed.